uncertainties/test_uncertainty_array.py

```diff
@@ -0,0 +1,276 @@
+    """Whether to box the data in a Series"""
+    return request.param
+
+
+@pytest.fixture
+def dtype():
+    return UncertaintyDtype()
+
+
+@pytest.fixture
+def data(request):
+    return UncertaintyArray(
+        [ufloat(i, abs(i)/100) for i in np.arange(start=1.0, stop=101.0)]
+    )
+
+
+@pytest.fixture
+def data_missing():
+    return UncertaintyArray(
+        [ufloat(i, abs(i)/100) for i in [np.nan, 1]]
+    )
+
+
+@pytest.fixture
+def data_for_twos():
+    x = [
+        2.0,
+    ] * 100
+    return UncertaintyArray(
+         [ufloat(i, abs(i)/100) for i in x]
+    )
+
+
+@pytest.fixture(params=["data", "data_missing"])
+def all_data(request, data, data_missing):
+    if request.param == "data":
+        return data
+    elif request.param == "data_missing":
+        return data_missing
+
+
+@pytest.fixture
+def data_repeated(data):
+    """Return different versions of data for count times"""
+
+    def gen(count):
+        for _ in range(count):
+            yield data
+
+    yield gen
+
+
+@pytest.fixture(params=[None, lambda x: x])
+def sort_by_key(request):
+    """
+    Simple fixture for testing keys in sorting methods.
+    Tests None (no key) and the identity key.
+    """
+    return request.param
+
+
+@pytest.fixture
+def data_for_sorting():
+    return UncertaintyArray(
+         [ufloat(i, abs(i)/100) for i in [0.3, 10.0, -50.0]]
+    )
+
+
+@pytest.fixture
+def data_missing_for_sorting():
+    return UncertaintyArray(
+         [ufloat(i, abs(i)/100) for i in [4.0, np.nan, -5.0]]
+    )
+
+
+@pytest.fixture
+def na_cmp():
+    """Binary operator for comparing NA values."""
+    return lambda x, y: pd.isna(x) and pd.isna(y)
+
+
+@pytest.fixture
+def na_value():
+    return pd.NA
+
+
+@pytest.fixture
+def data_for_grouping():
+    a = ufloat(1.0, 0.1)
+    b = ufloat(2.0, 0.1)
+    c = ufloat(3.0, 0.1)
+    x = [a, a, np.nan, np.nan, b, b, a, c]
+    return UncertaintyArray(x)
+
+# === missing from pandas extension docs about what has to be included in tests ===
+# copied from pandas/pandas/conftest.py
+_all_arithmetic_operators = [
+    "__add__",
+    "__radd__",
+    "__sub__",
+    "__rsub__",
+    "__mul__",
+    "__rmul__",
+    "__floordiv__",
+    "__rfloordiv__",
+    "__truediv__",
+    "__rtruediv__",
+    "__pow__",
+    "__rpow__",
+    "__mod__",
+    "__rmod__",
+]
+
+
+@pytest.fixture(params=_all_arithmetic_operators)
+def all_arithmetic_operators(request):
+    """
+    Fixture for dunder names for common arithmetic operations
+    """
+    return request.param
+
+
+@pytest.fixture(params=["__eq__", "__ne__", "__le__", "__lt__", "__ge__", "__gt__"])
+def all_compare_operators(request):
+    """
+    Fixture for dunder names for common compare operations
+
+    * >=
+    * >
+    * ==
+    * !=
+    * <
+    * <=
+    """
+    return request.param
+
+
+# commented functions aren't implemented in numpy/pandas
+_all_numeric_reductions = [
+    "sum",
+    "max",
+    "min",
+    "mean",
```

Comment:
`mean` is a non-trivail calculation in the world of uncertainties.  I don't think `uncertainties` does it.  I've created code that does it:

```
# https://stackoverflow.com/a/74137209/1291237                                                                                                                                                                                         
def umean(unquantified_data):
    """                                                                                                                                                                                                                                
    Assuming Gaussian statistics, uncertainties stem from Gaussian parent distributions. In such a case,                                                                                                                               
    it is standard to weight the measurements (nominal values) by the inverse variance.                                                                                                                                                
                                                                                                                                                                                                                                       
    Following the pattern of np.mean, this function is really nan_mean, meaning it calculates based on non-NaN values.                                                                                                                 
    If there are no such, it returns np.nan, just like np.mean does with an empty array.                                                                                                                                               
                                                                                                                                                                                                                                       
    This function uses error propagation on the to get an uncertainty of the weighted average.                                                                                                                                         
    :param: A set of uncertainty values                                                                                                                                                                                                
    :return: The weighted mean of the values, with a freshly calculated error term                                                                                                                                                     
    """
    arr = np.array(
	[v if isinstance(v, ITR.UFloat) else ITR.ufloat(v, 0) for v in unquantified_data if not ITR.isnan(v)]
    )
    N = len(arr)
    if N == 0:
        return np.nan
    if N == 1:
        return arr[0]
    nominals = ITR.nominal_values(arr)
    if any(ITR.std_devs(arr) == 0):
        # We cannot mix and match "perfect" measurements with uncertainties                                                                                                                                                            
        # Instead compute the mean and return the "standard error" as the uncertainty                                                                                                                                                  
        # e.g. ITR.umean([100, 200]) = 150 +/- 50                                                                                                                                                                                      
        w_mean = sum(nominals) / N
        w_std = np.std(nominals) / np.sqrt(N - 1)
    else:
        # Compute the "uncertainty of the weighted mean", which apparently                                                                                                                                                             
        # means ignoring whether or not there are large uncertainties                                                                                                                                                                  
        # that should be created by elements that disagree                                                                                                                                                                             
        # e.g. ITR.umean([100+/-1, 200+/-1]) = 150.0+/-0.7 (!)                                                                                                                                                                         
        w_sigma = 1 / sum([1 / (v.s**2) for v in arr])
        w_mean = sum([v.n / (v.s**2) for v in arr]) * w_sigma
        w_std = w_sigma * np.sqrt(sum([1 / (v.s**2) for v in arr]))
    result = ITR.ufloat(w_mean, w_std)
    return result
```

Reply:
This is the sort of thing that `__array_function__` would help with, so I could do `np.mean(UArray)` without needing to understand the uncertainty logic